openrc: adapt policy to use of seedrng

valasiadis opened the pull request:
openrc-0.45 changed the seeding mechanism for the random number generator [1]. The changed policy defines rules to enable accurate execution of seedrng at boot time.

[1] https://github.com/OpenRC/openrc/commit/076c2552aeff88a27fe275dfaae61dedf4bb4bd5

0xC0ncord commented:
Couple things:

1. Since it looks like seedrng can be used by other programs, I'm thinking this should go into a different policy module, probably `miscfiles`.
2. This should probably go upstream too, so we'll need to use interfaces to use these seedrng files instead of using `gen_require()`.

WDYT @perfinion ?

0xC0ncord commented:
OK it turns out there is already a type for this exact thing which already has the access needed.

It looks like all we need to do is label `/var/lib/seedrng` as `init_random_seed_t`.

valasiadis commented:
> OK it turns out there is already a type for this exact thing which already has the access needed.
> 
> It looks like all we need to do is label `/var/lib/seedrng` as `init_random_seed_t`.

Sounds perfect. The `init_manage_random_seed` interface (`system/init.if`) covers rw permissions for the `var_lib_t` and `var_run_t` dirs as well as `manage_file_perms` for the `init_random_seed_t`. The interface is only called for `initrc_t` and conditionally for `init_t`, which is fine. If the relabeling isn't a problem (i.e. the dir created by seedrng is on creation labeled according to the fc, not its parent directory), then I guess we're good just by adding a relabel context for the file.

Should I put the statement in `system/init.fc`?

valasiadis commented:
@0xC0ncord Anything?

0xC0ncord commented:
Sorry for the delay.

> Should I put the statement in system/init.fc?

Yes that should be all we need.

WavyEbuilder commented on policy/modules/contrib/openrc.te:
```diff
@@ -33,3 +33,17 @@ fs_manage_cgroup_files(openrc_cgroup_release_t)
+# to allow the relabeling of the files:
+allow setfiles_t seedrng_var_lib_t:dir { relabelto getattr read open search };
+allow setfiles_t seedrng_var_lib_t:file { relabelto getattr };
+allow seedrng_var_lib_t fs_t:filesystem { associate };
```
No need for braces for a single permission.

WavyEbuilder replied:
Generally it's useful as it is in line with our existing Style for reference policy.

WavyEbuilder commented on policy/modules/system/init.fc:
```diff
@@ -75,6 +75,9 @@ ifdef(`distro_redhat',`
 /var/lib/random-seed	--	gen_context(system_u:object_r:init_random_seed_t,s0)
 /var/lib/systemd(/.*)?		gen_context(system_u:object_r:init_var_lib_t,s0)
+ifdef(`distro_gentoo',`
+/var/lib/seedrng(/seed\.(no-)?credit)?	gen_context(system_u:object_r:init_random_seed_t)
```
As in line with the rest of our policy, please indent the filecon statement as it is inside a conditional macro.

WavyEbuilder replied:
Yes, however I think there are a few unintentional changes. `--` implies that you are labelling only files, but I'm guessing by the `(/seed\.(no-)?credit)?`, that `/var/lib/seedrng` is a directory (is that correct)? However, I don't see much benefit from not just using `(/.*)?` here; everything in `/var/lib/seedrng` should be `init_random_seed_t`, and being overly-specific with the labelling adds extra work in the future (if e.g. the specific files stored in `/var/lib/seedrng` changes), isn't particularly accurate (I'd argue that entire dir *and* it's contents are `init_random_seed_t`). Also, you're missing an MLS classification. Something like this should be good:
```
ifdef(`distro_gentoo',`
/var/lib/seedrng(/.*)?  gen_context(system_u:object_r:init_random_seed_t,s0)
')
```

valasiadis replied:
Changed it. Does that look okay to you? I force-pushed again, so you need to look at the current commit...

WavyEbuilder commented:
> i.e. the dir created by seedrng is on creation labeled according to the fc, not its parent directory

@0xC0ncord From what I can see, this is not the case. We need to do more than just labelling here as I cannot find an existing transition. A name-based filetransition should suffice from scontext `var_lib_t` to tcontext `init_random_seed_t` class `dir` name `seedrng`.

@valasiadis Would you mind adding this please?

0xC0ncord commented:
> @0xC0ncord From what I can see, this is not the case. We need to do more than just labelling here as I cannot find an existing transition. A name-based filetransition should suffice from scontext `var_lib_t` to tcontext `init_random_seed_t` class `dir` name `seedrng`.

Ah you're right. Glad you caught that!

valasiadis commented:
Sorry for the late reply, I totally forgot about this issue. I did some changes now, but am not able to test this at the moment. Perhaps you could have a look?

valasiadis commented on policy/modules/system/init.te:
```diff
@@ -1022,6 +1022,12 @@ ifdef(`distro_gentoo',`
 		dhcpd_setattr_state_files(initrc_t)
 	')
+	
+	# seedrng
```
I see. I have basically no good understanding of M4 or those SELinux policies. All the things I read four months ago, I already forget. So thanks for your guidance here :D
Please have a look at the recent changes ;)

WavyEbuilder replied:
> btw. `init_t` or `initrc_t`?

See below, probably `initrc_t as it runs as its own binary (`/usr/bin/seedrng`). I'll write a policy module to confine it into its own domain later.

WavyEbuilder commented on policy/modules/system/init.te:
```diff
@@ -1021,7 +1023,7 @@ ifdef(`distro_gentoo',`
 	optional_policy(`
 		dhcpd_setattr_state_files(initrc_t)
-	')
+	')	
```
Why the random whitespace change? Probably best to get rid of this :P

WavyEbuilder commented on policy/modules/system/init.te:
```diff
@@ -991,6 +991,8 @@ ifdef(`distro_gentoo',`
 	files_manage_generic_locks(initrc_t)
 	files_manage_var_symlinks(initrc_t)
 	files_runtime_filetrans(initrc_t, initrc_state_t, dir, "openrc")
+	# seedrng
```
Comment a bit unnecessary here imo

WavyEbuilder commented on policy/modules/system/init.te:
```diff
@@ -991,6 +991,8 @@ ifdef(`distro_gentoo',`
 	files_manage_var_symlinks(initrc_t)
 	files_runtime_filetrans(initrc_t, initrc_state_t, dir, "openrc")
+	# seedrng
+	files_var_lib_filetrans(init_t, init_random_seed_t, dir, "seedrng")
```
Well on closer look, this is *probably* `initrc_t`, so if you could change it to that that'd be great. Also add a TODO comment along these lines above please:
```
TODO: Give seenrng its own domain. It's provided by openrc, but other things can use it, and regardless it'd be good to confine it, even in early boot.
```

valasiadis replied:
Done. Looks good to you?

WavyEbuilder commented on policy/modules/system/init.te:
```diff
@@ -1016,6 +991,8 @@ ifdef(`distro_gentoo',`
 	files_manage_generic_locks(initrc_t)
 	files_manage_var_symlinks(initrc_t)
 	files_runtime_filetrans(initrc_t, initrc_state_t, dir, "openrc")
+	# TODO: Give seenrng its own domain. It's provided by openrc, but other things can use it, and regardless it'd be good to confine it, even in early boot.
```
Ah wait, this has a problem. Sadly, we can't use the `'` symbol in comments because of `m4`....

Could you update this to remove all instances of `'` (so basically just expand the contractions `It's` and `it'd`) please? I'll merge it in after.

Thanks!

valasiadis replied:
Done. Thanks for your support btw.

WavyEbuilder replied:
Not a problem at all - thank you for the patch! Merged :)

WavyEbuilder commented:
@valasiadis Could you force push it as a single commit? Also, your second commit is missing a DCO signoff: https://cert-manager.io/docs/contributing/sign-off/

valasiadis commented:
> @valasiadis Could you force push it as a single commit? Also, your second commit is missing a DCO signoff: https://cert-manager.io/docs/contributing/sign-off/

My bad...

valasiadis committed:
openrc: adapt policy to use of seedrng

openrc-0.45 changed the seeding mechanism for the random number generator [1]. The changed policy defines rules to enable accurate execution of seedrng at boot time.

[1] OpenRC/openrc@076c255

Signed-off-by: Emanuel Valasiadis <emanuel@valasiadis.space>

valasiadis commented:
I included another local change in that commit, sorry. New commit should be okay though
